src/core/utilities/bucketlist.hpp

```diff
@@ -0,0 +1,315 @@
+};
+
+
+
+template<std::size_t bucket_size>
+class BucketList
+{
+    // ----------------- BucketList Iterator ----------------------------------
+    template<typename BucketListRef>
+    class bucket_iterator : public std::iterator<std::forward_iterator_tag, std::size_t>
+    {
+    private:
+        using Bindex = typename BucketListIndex::Bindex;
+
+    public:
+        auto& operator*() const { return bucketsList_.buckets_[curr_bucket_][curr_pos_]; }
+        auto& operator*() { return bucketsList_.buckets_[curr_bucket_][curr_pos_]; }
+        auto operator++();
+        bool operator!=(bucket_iterator const& other) const;
+
+        bucket_iterator(BucketListRef blist, Bindex curr_bucket = 0, Bindex curr_pos = 0)
+            : curr_bucket_{curr_bucket}
+            , curr_pos_{curr_pos}
+            , bucketsList_{blist}
+        {
+        }
+
+    private:
+        typename BucketListIndex::Bindex curr_bucket_ = 0, curr_pos_ = 0;
+        BucketListRef bucketsList_;
+    };
+    // ----------------- END BucketList Iterator ------------------------------
+
+
+
+
+public:
+    BucketList(std::size_t min_bucket_nbr = 1)
+        : buckets_(min_bucket_nbr)
+    {
+        for (auto& b : buckets_)
+            for (auto& v : b)
+                v = BucketListIndex::default_idx_v;
+    }
+
+    using iterator       = bucket_iterator<BucketList&>;
+    using const_iterator = bucket_iterator<BucketList const&>;
+
+    void add(std::size_t itemIndex);
+    void remove(std::size_t itemIndex);
+    bool in_bucket(std::size_t itemIndex);
+
+    // reallocate bucketlist memory if more empty space than max_empty
+    void trim(std::size_t max_empty);
+
+    // to use if an item in an indexed array is moved at another index
+    void updateIndex(std::size_t oldIndex, std::size_t newIndex);
+
+    // empty the bucketlist, but leaves the capacity untouched
+    void empty();
+    bool is_empty() const { return bucket_idx == 0 and curr == 0; }
+
+    std::size_t size() const { return bucket_size * (bucket_idx) + curr; }
+    std::size_t capacity() const { return buckets_.capacity() * bucket_size; }
+
+    auto begin() { return iterator{*this}; }
+    auto begin() const { return const_iterator{*this}; }
+    auto cbegin() const { return const_iterator{*this}; }
+    auto end();
+    auto end() const;
+    auto cend() const;
+
+
+
+private:
+    BucketListIndex lastIdx_() const;
+
+
+    void decrement_()
+    {
+        if (curr == 0)
+        {
+            bucket_idx--;
+            curr = bucket_size - 1;
+        }
+        else
+        {
+            auto check
+                = bucket_idx >= 0
+                  and bucket_idx < static_cast<typename BucketListIndex::Bindex>(buckets_.size())
+                  and curr >= 0 and (curr - 1 < static_cast<BucketListIndex::Bindex>(bucket_size));
+            if (!check)
+            {
+                std::cout << curr << " " << bucket_idx << "\n";
+            }
+            assert(check);
```

Comment:
should be throw?
assert don't trigger in  release build

Reply:
yeah, I was not sure this check should remain.... I'll leave it at least for some time... we'll see if that impacts perfs... and transform it to an exception